torch/distributed/device_mesh.py

```diff
@@ -65,84 +67,41 @@ def get_current_mesh(self) -> "DeviceMesh":
-                    # then the final slice_dim_idx should be [0, 1, 2].
-                    slice_dim_idx.append(mesh_dim_indices[0] - num_dims_flatten)
-                    num_dims_flatten += len(mesh_dim_indices) - 1
+            for name in submesh_dim_names:
+                if name in not_none(device_mesh.mesh_dim_names):
                     slice_dim_group_name.append(
-                        self.root_to_flatten_mapping[device_mesh][
-                            mesh_dim_name
-                        ]._dim_group_names[0]  # type: ignore[has-type]
+                        device_mesh._dim_group_names[  # type: ignore[has-type]
+                            not_none(device_mesh.mesh_dim_names).index(name)
+                        ]
                     )
                 else:
-                    slice_dim_idx.append(mesh_dim_indices[0] - num_dims_flatten)
+                    flatten_mesh = self.root_to_flatten_mapping[device_mesh][name]
                     slice_dim_group_name.append(
-                        device_mesh._dim_group_names[mesh_dim_indices[0]]  # type: ignore[has-type]
-                    )
-
-            # mesh_tensor has already been flattened if needed. So mesh_tensor.ndim <= device_mesh.mesh.ndim now.
-            mesh_dims_remained_idx = list(range(mesh_tensor.ndim))
-            for idx in slice_dim_idx:
-                if idx not in mesh_dims_remained_idx:
-                    raise NotImplementedError(
-                        "Currently, this only allows slicing out a contiguous flattened dim."
+                        flatten_mesh._dim_group_names[  # type: ignore[has-type]
+                            not_none(flatten_mesh.mesh_dim_names).index(name)
+                        ]
                     )
-                mesh_dims_remained_idx.remove(idx)
-
-            # pg_ranks_by_dim is the size of [number of local ranks of the outermost submesh dimension, *slice_dim_idx]
-            # This means on each local rank of the outermost slice mesh dim, we have a tensor of submesh size with
-            # the pg ranks of the submesh. From this, we can extract the submesh mesh tensor contains the current rank.
-            pg_ranks_by_dim = mesh_tensor.permute(
-                *mesh_dims_remained_idx, *slice_dim_idx
-            ).reshape(-1, *slice_dim_size)
-
             cur_rank = device_mesh.get_rank()
+            pg_ranks_by_dim = layout.to_remapping_tensor(
+                device_mesh.mesh,
+                get_world_size(),
+            )
             for mesh_nd in pg_ranks_by_dim:
                 submesh = DeviceMesh(
                     device_mesh.device_type,
                     mesh_nd,
                     mesh_dim_names=submesh_dim_names,
                     _init_backend=False,
+                    _layout=layout,
                 )
                 if cur_rank in mesh_nd:
                     res_submesh = submesh
```

Comment:
Can you explain this code? I know that it wasn't added by this PR, but I find it odd. Basically, we create a potentially very large number of DeviceMesh objects (e.g., on 1024 GPUs with TP=2, this could be up to 512 DeviceMeshes), only to throw them all away except for one. Why do we do that?

I suspect this means that constructing a DeviceMesh objects must have some global side-effects. Is that so? What kind of effects are we talking about?

Reply:
Note that in this snipped `init_backend` is False so we're not even creating sub-PGs...